src/dune/lib_dep.ml

```diff
@@ -1,32 +1,100 @@
 open Stdune
 
 module Select = struct
-  type choice =
-    { required : Lib_name.Set.t
-    ; forbidden : Lib_name.Set.t
-    ; file : string
-    }
+  module Choice = struct
+    type t =
+      { required : Lib_name.Set.t
+      ; forbidden : Lib_name.Set.t
+      ; file : string
+      }
 
-  let dyn_of_choice { required; forbidden; file } =
-    let open Dyn.Encoder in
-    record
-      [ ("required", Lib_name.Set.to_dyn required)
-      ; ("forbidden", Lib_name.Set.to_dyn forbidden)
-      ; ("file", string file)
-      ]
+    let decode ~result_fn =
+      let open Dune_lang.Decoder in
+      enter
+        (let* dune_version = Dune_lang.Syntax.get_exn Stanza.syntax in
+         let+ loc = loc
+         and+ preds, file =
+           until_keyword "->"
+             ~before:
+               (let+ s = string
+                and+ loc = loc in
+                let loc = Some loc in
+                match String.drop_prefix s ~prefix:"!" with
+                | Some s -> Right (Lib_name.of_string_exn ~loc s)
+                | None -> Left (Lib_name.of_string_exn ~loc s))
+             ~after:(located filename)
+         in
+         match file with
+         | None ->
+           User_error.raise ~loc
+             [ Pp.textf "(<[!]libraries>... -> <file>) expected" ]
+         | Some (loc_file, file) ->
+           let () =
+             if dune_version >= (2, 0) then (
+               let prefix, suffix =
+                 match String.rsplit2 result_fn ~on:'.' with
```

Comment:
What about using `Filename.split_extension`? so that the selector appears before the extension

Reply:
I think the select has to appear before the extension here. See the tests for example. Do you have a test case perhaps?

Reply:
Yh it's just for things like `foo.cppo.ml`. The extension of `foo.cppo.ml` is `.cppo.ml` while `rsplit2` will return `.ml`

Reply:
Sure, but this might not necessarily what the user might want. For example, he might have the 2 different sources preprocessed differently. For example, there might be the file `foo.cppo.ml` but with sources `foo.bar.ml` and `foo.baz.cppo.ml`

Reply:
hmm, but with the current code, don't the choices have to be of the form `foo.cppo.XXX.ml`?

Reply:
and we would instead like for them to be of the form `foo.XXX.cppo.ml`

Reply:
@diml okay, I switched to using `Filename.split_extension`.